Source/bmalloc/bmalloc/PerThread.h

```diff
@@ -92,6 +97,33 @@ template<> struct PerThreadStorage<PerHeapKind<Cache>> {
 #pragma clang diagnostic pop
 };
 
+#elif BPLATFORM(WIN)
+
+template<typename T> class PerThreadStorage {
+public:
+    ~PerThreadStorage()
+    {
+        if (m_destructor && m_object)
+            m_destructor(m_object);
+    }
+
+    static void* get() { return s_storage.m_object; }
+
+    static void init(void* object, void (*destructor)(void*))
+    {
+        s_storage.m_object = object;
+        s_storage.m_destructor = destructor;
+    }
+
+private:
+    static thread_local PerThreadStorage<T> s_storage;
+
+    void* m_object;
+    void (*m_destructor)(void*);
+};
+
+template<typename T> thread_local PerThreadStorage<T> PerThreadStorage<T>::s_storage;
```

Comment:
I think the question is how this `thread_local` works. It is possible that,

1. One thread have multiple `thread_local`s and now thread is dying
2. This bmalloc one gets destroyed
3. Then, different `thread_local` gets destroyed.
4. And (3) invokes `bmalloc::free`, and it touches this bmalloc's `thread_local`

So the question is what happens in the above case (it would be super common case)?

Right now, POSIX behaviors of pthread's thread_local is, whenever it is set again, then thread_local acts again. And then, the destruction of this thread_local will be called later again, and it will be done N times. (This N is `PTHREAD_DESTRUCTOR_ITERATIONS`). So this revival is fine for POSIX. But we need to know how Windows behaves too. If it behaves similarly then it is fine.

Reply:
Probably, Windows TLS is not working in this way. So we need much more complicated solution here.
Probably, the solution would be similar to what libpas is doing for non-Darwin.

1. Add plain pointer TLS. `__declspec(thread)` And pointing it to `NULL` via static initialization.
2. We init per-thread Cache and set this pointer to this field only when this field was `NULL`. Otherwise, fallback to the global implementation.
3. Register thread destructor via Windows CRT section data
4. destructor destroys (2)'s heap, and set special value like, `1` to this pointer to prevent from recursive initialization of this.

Reply:
You probably need to define `DllMain` and register it. Then you can observe `DLL_THREAD_DETACH`.

Reply:
Thanks for the comment. This is why I open this thread PR separately from other part. I've kept feeling something wrong in this implementation and wanted to hear from experts. I'll think carefully about it and revisit here again soon.